src/setupWorker/start/utils/activateMocking.ts

```diff
@@ -1,15 +1,23 @@
 import { addMessageListener } from '../../../utils/createBroadcastChannel'
 import { StartOptions } from '../../glossary'
+import { MSWEventListener } from '../../../utils/internal/mswEventListener'
 
 export const activateMocking = (
   worker: ServiceWorker,
   options?: StartOptions,
 ) => {
+  let listeners: MSWEventListener[]
+  const removeLiteners = () => {
+    listeners.forEach((listener) =>
+      listener.handler.removeEventListener(listener.type, listener.listener),
+    )
+  }
+
```

Comment:
I have done in this way because activateMocking listeners are removed asap, when the promise is resolved. So I thought to not add to context listeners

Reply:
Even in that case, I think it's best to still put that onto the context in so that it can re-use the global teardown in the event that the window is torn down, or stop is called before the immediate message is received.

For example, I believe there is a race condition that will happen if:

1. `addMessagelistener()` is called
2. user calls `stop()`
3. message is received.

In that case, the listener will still be dispatched when it shouldn't be.

I think we can have the best of both worlds by implementing a `once()` method on the context, similar to `EventEmitter#once` which lets us have the best of both worlds. It removes itself immediately, but it also uses the context-bound event registration mechanism that won't overstay its welcome.

Here is an example implementation https://github.com/cowboyd/msw/blob/82fa2ad486f0030e6e4fb9f352f295cedb45de92/src/setupWorker/setupWorker.ts#L62

As a bonus, I made it return a promise so that you can now use simple `async` functions to work with message dispatch and response https://github.com/cowboyd/msw/blob/82fa2ad486f0030e6e4fb9f352f295cedb45de92/src/setupWorker/setupWorker.ts#L62-L100

Where this approach really shines is in things like the integrity check which can now be written as a simple `async` function. Thus this:

```ts
export function requestIntegrityCheck(
  serviceWorker: ServiceWorker,
): Promise<ServiceWorker> {
  let listeners: MSWEventListener[]
  // Signal Service Worker to report back its integrity
  serviceWorker.postMessage('INTEGRITY_CHECK_REQUEST')
  const removeLiteners = () => {
    listeners.forEach((listener) =>
      listener.handler.removeEventListener(listener.type, listener.listener),
    )
  }
  return new Promise((resolve, reject) => {
    listeners = addMessageListener(
      'INTEGRITY_CHECK_RESPONSE',
      (message) => {
        const { payload: actualChecksum } = message

        // Compare the response from the Service Worker and the
        // global variable set by webpack upon build.
        if (actualChecksum !== SERVICE_WORKER_CHECKSUM) {
          return reject(
            new Error(
              `Currently active Service Worker (${actualChecksum}) is behind the latest published one (${SERVICE_WORKER_CHECKSUM}).`,
            ),
          )
        }
        removeLiteners()
        resolve(serviceWorker)
      },
      () => {
        removeLiteners()
        reject()
      },
    )
  })
}
```

becomes this:

```ts
export async function requestIntegrityCheck(
  context: SetupWorkerInternalContext,
  serviceWorker: ServiceWorker,
): Promise<ServiceWorker> {
  // Signal Service Worker to report back its integrity
  serviceWorker.postMessage('INTEGRITY_CHECK_REQUEST')

  const { payload: actualChecksum } = await context.once('INTEGRITY_CHECK_RESPONSE')

  // Compare the response from the Service Worker and the
  // global variable set by webpack upon build.
  if (actualChecksum !== SERVICE_WORKER_CHECKSUM) {
    throw new Error(
      `Currently active Service Worker (${actualChecksum}) is behind the latest published one (${SERVICE_WORKER_CHECKSUM}).`,
    )
  }
  return serviceWorker
}
```

Which seems like a pretty nice win.